filament/backend/test/SharedShaders.cpp

```diff
@@ -0,0 +1,257 @@
+    gl_Position = vec4(
+            mesh_position.xy * (params.scaleMinusOne.xy + 1.0) + params.offset.xy,
+            params.scaleMinusOne.z + 1.0,
+            1.0);
+#if defined(TARGET_VULKAN_ENVIRONMENT)
+    // In Vulkan, clip space is Y-down. In OpenGL and Metal, clip space is Y-up.
+    gl_Position.y = -gl_Position.y;
+#endif
+})" };
+        }
+        case VertexShaderType::Textured: {
+            return ShaderText{
+                    R"(
+#version 450 core
+layout(location = 0) in vec4 mesh_position;
+layout(location = 0) out vec2 uv;
+)", R"(
+void main() {
+    gl_Position = vec4(mesh_position.xy, 0.0, 1.0);
+    uv = (mesh_position.xy * 0.5 + 0.5);
+#if defined(TARGET_VULKAN_ENVIRONMENT)
+    // In Vulkan, clip space is Y-down. In OpenGL and Metal, clip space is Y-up.
+    gl_Position.y = -gl_Position.y;
+#endif
+})" };
+        }
+        default:
+            return std::nullopt;
+    }
+}
+
+std::optional<ShaderText> GetGlslFragmentShader(FragmentShaderType type) {
+    switch (type) {
+        case FragmentShaderType::White: {
+            return ShaderText{
+                    R"(
+#version 450 core
+precision mediump int; precision highp float;
+layout(location = 0) out vec4 fragColor;
+)", R"(
+void main() {
+    fragColor = vec4(1.0, 1.0, 1.0, 1.0);
+})" };
+        }
+        case FragmentShaderType::SolidColored: {
+            return ShaderText{
+                    R"(
+#version 450 core
+precision mediump int; precision highp float;
+layout(location = 0) out vec4 fragColor;
+)", R"(
+void main() {
+    fragColor = params.color;
+})" };
+        }
+        case FragmentShaderType::Textured: {
+            return ShaderText{
+                    R"(
+#version 450 core
+precision mediump int; precision highp float;
+layout(location = 0) out vec4 fragColor;
+layout(location = 0) in vec2 uv;
+
+layout(location = 0, set = 1) uniform sampler2D test_tex;
+
+void main() {
+    fragColor = texture(test_tex, uv);
+})" };
+        }
+        default:
+            return std::nullopt;
+    }
+}
+
+std::optional<std::string> GetGlslUniform(ShaderUniformType type) {
+    switch (type) {
+        case ShaderUniformType::None: {
+            return "";
+        }
+        case ShaderUniformType::Simple: {
+            return R"(
+layout(binding = 0, set = 1) uniform Params {
+    highp vec4 color;
+    // Use scaleMinusOne instead of scale so that a 0 initialized value is a good default
+    highp vec4 scaleMinusOne;
+    highp vec4 offset;
+} params;
+)";
+        }
+        case ShaderUniformType::SimpleWithPadding: {
+            return R"(
+layout(binding = 0, set = 1) uniform Params {
+    highp vec4 padding[4];  // offset of 64 bytes
+
+    highp vec4 color;
+    // Use scaleMinusOne instead of scale so that a 0 initialized value is a good default
+    highp vec4 scaleMinusOne;
+    highp vec4 offset;
+} params;
+)";
+        }
+        case ShaderUniformType::Sampler: {
+            return "";
+            /*return R"(
```

Comment:
Remove?

Reply:
Whoops, thanks this was in progress. I needed to move the uniform definition out of the fragment shader and into here.